File: TouchControl/src/io/github/mattson543/touchcontrol/touchables/ToggleButton.java
package io.github.mattson543.touchcontrol.touchables;

import org.opencv.core.*;

/**
 * Abstract Button - contains general (toggle) button behavior.
 *
 * @author mattson543
 */
public abstract class ToggleButton extends Button
{
	/**
	 * The current state of the button
	 */
	protected boolean isToggledOn;
	/**
	 * Whether or not the button has already switched states internally
	 */
	private boolean hasSwitched;

	protected ToggleButton(Rect dimensions, Scalar color)
	{
		super(dimensions, color);
		isToggledOn = false;
		hasSwitched = false;
	}

	/*
	 * (non-Javadoc)
	 * @see
	 * io.github.mattson543.touchcontrol.touchables.Button
	 * #updateDetectionPoint(org.opencv.core.Mat)
	 */
	@Override
	public Point updateDetectionPoint(Mat filteredImage)
	{
		//Save previous state
		boolean wasClicked = isBeingClicked;

		//Update state
		super.updateDetectionPoint(filteredImage);

		//Toggle state
		if (isBeingClicked && !wasClicked)
		{
			if (!hasSwitched)
			{
				isToggledOn = !isToggledOn;
				hasSwitched = true;
			}
		}
		else
			hasSwitched = false;

		return detectionPoint;
	}

	/*
	 * (non-Javadoc)
	 * @see io.github.mattson543.touchcontrol.touchables.Button#toString()
	 */
	@Override
	public String toString()
	{
		return super.toString() + format("Toggled on:", isToggledOn);
	}
}

File: src/main/java/io/github/jakejmattson/touchcontrol/touchables/ToggleButton.kt
package io.github.jakejmattson.touchcontrol.touchables

import org.opencv.core.*

/**
 * Abstract Button - contains general (toggle) button behavior.
 *
 * @author JakeJMattson
 */
abstract class ToggleButton protected constructor(dimensions: Rect, color: Scalar): Button(dimensions, color) {
	/**
	 * The current state of the button
	 */
	protected var isToggledOn: Boolean = false
	/**
	 * Whether or not the button has already switched states internally
	 */
	private var hasSwitched: Boolean = false

	override fun updateDetectionPoint(filteredImage: Mat): Point? {
		//Save previous state
		val wasClicked = isBeingClicked

		//Update state
		super.updateDetectionPoint(filteredImage)

		//Toggle state
		if (isBeingClicked && !wasClicked) {
			if (!hasSwitched) {
				isToggledOn = !isToggledOn
				hasSwitched = true
			}
		}
		else
			hasSwitched = false

		return detectionPoint
	}

	override fun toString() = super.toString() + format("Toggled on:", isToggledOn)
}